requests-for-comments/2024-01-23-project-environments.md

```diff
@@ -0,0 +1,116 @@
+# Request for comments: Project environments, by Michael Matloka
+## Context
+
+
+Our users commonly operate multiple environments of their product. The two common types of split by environment are:
+
+1. By stage, driven by the engineering workflow, e.g. dev/staging/production.
+2. By region, driven by data protection requirements, e.g. country A/country B.
+
+When doing that, customers typically want to share setup between environments, while ensuring *data* from one environment doesn't pollute the others.
+
+The multiple environments use case is one of the reasons why projects are a feature of PostHog, and projects are great for tracking separate products, but not ideal for tracking multiple instances of the same product: whatever you set up for one project, you have to manually replicate in all others.
+
+We currently solve this in Feature flags with a feature for syncing flags between projects, based on the flag key. Feature flags make this uniquely easy thanks to the feature flag key being an effective organization-wide identifier. No other entity – dashboards, insights, etc. – has this advantage though.
+
+Because feature flags use cohorts and actions in their definitions, these are also synced in the current solution –but only based on the cohort/action name, so if the name is changed in one environment, the link is silently broken.
```

Comment:
Hmm, this isn't right. When the name is changed in the target environment, we create a new cohort/action that is the same name as the source environment.

So nothing should ever be silently broken, it's just that _new_ synced flags will connect not to a cohort you've changed in the target environment, but a copy of the cohort as it currently is in the source environment.

Reply:
Yup, that's what I mean by the _link_ being silently broken. The feature flag itself is not broken, but the syncing mechanism is very implicit, so there's no visibility into how the dependencies of a flag are synced on subsequent syncs.